File: environment/src/main/java/jetbrains/exodus/tree/ITreeCursor.java
/**
 * Copyright 2010 - 2018 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree;

import jetbrains.exodus.ByteIterable;
import jetbrains.exodus.env.Cursor;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface ITreeCursor extends Cursor {

    boolean isMutable();

    ITree getTree();

    ITreeCursor EMPTY_CURSOR = new ITreeCursor() {

        @Override
        public boolean isMutable() {
            return false;
        }

        @Override
        public boolean getNext() {
            return false;
        }

        @Override
        public boolean getNextDup() {
            return false;
        }

        @Override
        public boolean getNextNoDup() {
            return false;
        }

        @Override
        public boolean getLast() {
            return false;
        }

        @Override
        public boolean getPrev() {
            return false;
        }

        @Override
        public boolean getPrevDup() {
            return false;
        }

        @Override
        public boolean getPrevNoDup() {
            return false;
        }

        @Override
        @NotNull
        public ByteIterable getKey() {
            throw new UnsupportedOperationException("No key found");
        }

        @Override
        @NotNull
        public ByteIterable getValue() {
            throw new UnsupportedOperationException("No value found");
        }

        @Override
        @Nullable
        public ByteIterable getSearchKey(@NotNull ByteIterable key) {
            return null;
        }

        @Override
        @Nullable
        public ByteIterable getSearchKeyRange(@NotNull ByteIterable key) {
            return null;
        }

        @Override
        public boolean getSearchBoth(@NotNull ByteIterable key, @NotNull ByteIterable value) {
            return false;
        }

        @Override
        @Nullable
        public ByteIterable getSearchBothRange(@NotNull ByteIterable key, @NotNull ByteIterable value) {
            return null;
        }

        @Override
        public int count() {
            return 0;
        }

        @Override
        public void close() {
        }

        @Override
        public ITree getTree() {
            return null;
        }

        @Override
        public boolean deleteCurrent() {
            return false;
        }
    };
}

File: environment/src/main/kotlin/jetbrains/exodus/tree/ITreeCursor.kt
/*
 * Copyright 2010 - 2023 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jetbrains.exodus.tree

import jetbrains.exodus.ByteIterable
import jetbrains.exodus.env.Cursor

interface ITreeCursor : Cursor {
    override fun isMutable(): Boolean
    val tree: ITree?

    companion object {
        @JvmField
        val EMPTY_CURSOR: ITreeCursor = object : ITreeCursor {
            override fun isMutable(): Boolean {
                return false
            }

            override fun getNext(): Boolean {
                return false
            }

            override fun getNextDup(): Boolean {
                return false
            }

            override fun getNextNoDup(): Boolean {
                return false
            }

            override fun getLast(): Boolean {
                return false
            }

            override fun getPrev(): Boolean {
                return false
            }

            override fun getPrevDup(): Boolean {
                return false
            }

            override fun getPrevNoDup(): Boolean {
                return false
            }

            override fun getKey(): ByteIterable {
                throw UnsupportedOperationException("No key found")
            }

            override fun getValue(): ByteIterable {
                throw UnsupportedOperationException("No value found")
            }

            override fun getSearchKey(key: ByteIterable): ByteIterable? {
                return null
            }

            override fun getSearchKeyRange(key: ByteIterable): ByteIterable? {
                return null
            }

            override fun getSearchBoth(key: ByteIterable, value: ByteIterable): Boolean {
                return false
            }

            override fun getSearchBothRange(key: ByteIterable, value: ByteIterable): ByteIterable? {
                return null
            }

            override fun count(): Int {
                return 0
            }

            override fun close() {}
            override val tree: ITree?
                get() = null

            override fun deleteCurrent(): Boolean {
                return false
            }
        }
    }
}
